crates/rustc_codegen_spirv/src/builder/load_store.rs

```diff
@@ -0,0 +1,506 @@
+                    .unwrap()
+                    .with_type(uint_ty);
+                uint_values_and_offsets.push((base_offset, lower));
+
+                let const_32 = self.constant_int(uint_ty, 32).def(self);
+                let shifted = self
+                    .emit()
+                    .shift_right_logical(ulong_ty, None, ulong_data, const_32)
+                    .unwrap();
+                let upper = self
+                    .emit()
+                    .u_convert(uint_ty, None, shifted)
+                    .unwrap()
+                    .with_type(uint_ty);
+                uint_values_and_offsets.push((base_offset + 1, upper));
+            }
+            _ => {
+                let mut err = self
+                    .tcx
+                    .sess
+                    .struct_err("Unsupported integer type for `codegen_internal_buffer_store`");
+                err.note(&format!("bits: `{:?}`", bits));
+                err.note(&format!("signed: `{:?}`", signed));
+                err.emit();
+            }
+        }
+    }
+
+    pub(crate) fn codegen_internal_buffer_store(
+        &mut self,
+        result_type: Word,
+        args: &[SpirvValue],
+    ) -> SpirvValue {
+        let uint_ty = SpirvType::Integer(32, false).def(rustc_span::DUMMY_SP, self);
+
+        let uniform_uint_ptr =
+            SpirvType::Pointer { pointee: uint_ty }.def(rustc_span::DUMMY_SP, self);
+
+        let zero = self.constant_int(uint_ty, 0).def(self);
+
+        let sets = self.bindless_descriptor_sets.borrow().unwrap();
+
+        let mut data = self.lookup_type(args[2].ty);
+
+        let bindless_idx = args[0].def(self);
+        let offset_arg = args[1].def(self);
+
+        let two = self.constant_int(uint_ty, 2).def(self);
+
+        let dword_offset = self
+            .emit()
+            .shift_right_arithmetic(uint_ty, None, offset_arg, two)
+            .unwrap();
+
+        let mut uint_values_and_offsets = vec![];
+        self.recurse_adt_for_stores(uint_ty, args[2], 0, &mut uint_values_and_offsets);
+
+        for (offset, uint_value) in uint_values_and_offsets {
+            let offset = if offset > 0 {
+                let element_offset = self.constant_int(uint_ty, offset as u64).def(self);
+
+                self.emit()
+                    .i_add(uint_ty, None, dword_offset, element_offset)
+                    .unwrap()
+            } else {
+                dword_offset
+            };
+
+            let indices = vec![bindless_idx, zero, offset];
+
+            let access_chain = self
+                .emit()
+                .access_chain(uniform_uint_ptr, None, sets.buffers, indices)
+                .unwrap()
+                .with_type(uniform_uint_ptr);
+
+            self.store(uint_value, access_chain, Align::from_bytes(0).unwrap());
+        }
+
+        self.emit_global()
+            .type_void()
+            .with_type(SpirvType::Void.def(rustc_span::DUMMY_SP, self))
```

Comment:
This... doesn't do what I think you think it does. I'm really confused if this passes spirv-val, it shouldn't, so I'm worried about test coverage if tests pass.

Reply:
Yeah it's quite an odd line, and I'm not sure if it's correct, so I could use some input here.

Reply:
I've pushed a change that addresses this - could you double check that it matches what we've discussed in our call?

Reply:
Yep, LGTM!